Core/Src/main.c

```diff
@@ -1,181 +1,249 @@
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+  uint8_t spi_transmit_data_buf, spi_receive_data_buf = 0;
+  // Need 10 bits to represent a reading
+  uint16_t spi_reading = 0;
+  const uint32_t adc_read_timeout = 5; // In ms
+
+  // Set PWM to minimum duty-cycle
+  __HAL_TIM_SET_COMPARE(&htim2, TIM_CHANNEL_1, ADC_MIN_COUNTS_DUTY_CYCLE);
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);   // Set CS to HIGH
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM2_Init();
+  /* USER CODE BEGIN 2 */
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);  // Start the PWM signal
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1) {
+	  loop_start: HAL_Delay(10);
+	  // Note size is in bits
+
+
+	  // --- Take a reading ---
+	  // Set CS to LOW
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);  
+	  // (1) pipe in 0b0000 0001
+	  spi_transmit_data_buf = 0x01;
+	  HAL_SPI_Transmit(&hspi1, &spi_transmit_data_buf,
+			           sizeof(uint8_t), adc_read_timeout);
+
+	  // (2) pipe in/out 0b1000 0000 / 0b0000 00RR
+	  spi_transmit_data_buf = 0xF0;
+	  HAL_SPI_TransmitReceive(&hspi1, &spi_transmit_data_buf,
+			  	  	  	  	  	  	  &spi_receive_data_buf,
+									  sizeof(uint16_t), adc_read_timeout);
+
+	  // Check we read in data properly
+	  if (spi_receive_data_buf > 3) {
+		  // We read data incorrectly, try to reset
+		  goto loop_start;
+	  }
+
+	  spi_reading ^= spi_receive_data_buf;
+	  spi_reading <<= 8; // Shift in in bits 9 and 8
+	  // (3) Grab remaining 8 bits from SPI
+	  spi_receive_data_buf = 0;
+	  HAL_SPI_Receive(&hspi1, &spi_receive_data_buf,
+			          sizeof(uint8_t), adc_read_timeout);
+
+	  spi_reading ^= spi_receive_data_buf;
+	  // Set CS to HIGH, reading is done
+	  HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+	  // Reset both buffers
+	  spi_transmit_data_buf = 0;
+	  spi_transmit_data_buf = 0;
+	  // --- Reading is Done ---
+	  // --- Set Timer Based on ADC Reading ---
+	  	  // 0 to 3.3v goes into adc
+	  	  // 0 to 1024 gets mapped to that range
+	  // Timer counts up to 960000 over 20ms
+	  // Max counts is (10%) * (960000) = 96000
+	  // Min counts is (5%) * (960000) = 48000
+
+	  __HAL_TIM_SET_COMPARE(&htim2, TIM_CHANNEL_1,
+			  	  	  	  	(uint32_t) ADC_VAL_TO_COUNTS * spi_reading + ADC_MIN_COUNTS_DUTY_CYCLE);

```

Comment:
calculation looks good! Although, make sure that your values aren't getting messed up because you are multiplying uints with doubles!

Reply:
Yeah upon further inspection looks like I don't even have to cast it to an int, I thought the macro required the input to be an int, I'll just omit the cast